target/riscv/cpu_helper.c

```diff
@@ -1447,16 +1447,22 @@ void riscv_cpu_do_interrupt(CPUState *cs)
 }
 
 #ifdef TARGET_CHERI
-void update_special_register_offset(CPURISCVState *env, cap_register_t *scr,
-                                    const char *name, target_ulong new_value)
+void update_special_register(CPURISCVState *env, cap_register_t *scr,
+                             const char *name, target_ulong new_value)
 {
-    // Any write to the CSR shall set the offset of the SCR to the value
-    // written. This shall be equivalent to a CSetOffset instruction, but with
-    // any exception condition instead just clearing the tag of the SCR.
-    target_ulong cursor = cap_get_cursor(scr);
-    target_ulong current_offset = cap_get_offset(scr);
-    target_ulong diff = new_value - current_offset;
-    target_ulong new_cursor = cursor + diff;
+    /* The new behaviour is that SCR updates affect the address. */
+    target_ulong new_cursor = new_value;
+    if (!CHERI_NO_RELOCATION(env)) {
+        /*
+         * Legacy behaviour: with relocation enabled, updates to the SCRs update
+         * the offset since updates to the architectural PC affect offset:
+         * This shall be equivalent to a CSetOffset instruction, but with any
+         * exception condition instead just clearing the tag of the SCR.
+         */
+        target_ulong current_offset = cap_get_offset(scr);
+        target_ulong diff = new_value - current_offset;
+        target_ulong new_cursor = cap_get_cursor(scr) + diff;
```

Comment:
```suggestion
        new_cursor = cap_get_cursor(scr) + diff;
```

Reply:
I just discovered this after spending 20 minutes debugging an untagged STCC... Should have just waited for your review :) Or turned on -Wshadow...

Reply:
CI should fail with a more sensible error message now.